### PR TITLE
GH-119258: Handle `STORE_ATTR_WITH_HINT` in tier two

### PR DESCRIPTION
**Pull request by brandtbucher**
This way we can remove the version check in the optimizer, too!

* Issue: gh-119258

### CODE REVIEW

File: Include/internal/pycore_dict.h [hunk 1]
```diff
@@ -246,7 +246,7 @@ dict_next_version(PyInterpreterState *interp)
     ((INTERP)->dict_state.global_version += DICT_VERSION_INCREMENT)
 #endif
 
-void
+PyAPI_FUNC(void)
```
Comment (markshannon) [inline Include/internal/pycore_dict.h:249]: ~Why does this symbol need to be exposed?
It isn't used by the JIT.~

It is used, via the `_PyDict_NotifyEvent` inline function.

File: Python/bytecodes.c [hunk 2]
```diff
@@ -2135,10 +2135,8 @@ dummy_func(
             _GUARD_DORV_NO_DICT +
             _STORE_ATTR_INSTANCE_VALUE;
 
-        inst(STORE_ATTR_WITH_HINT, (unused/1, type_version/2, hint/1, value, owner --)) {
+        op(_STORE_ATTR_WITH_HINT, (hint/1, value, owner --)) {
             PyTypeObject *tp = Py_TYPE(owner);
```
Comment (markshannon) [inline Python/bytecodes.c]:
```suggestion
```

This is only used in the assert, AFAICT, so will generate a compiler warning for non-debug builds.

File: Python/bytecodes.c [hunk 3]
```diff
@@ -2135,10 +2135,8 @@ dummy_func(
...
-            assert(type_version != 0);
-            DEOPT_IF(tp->tp_version_tag != type_version);
             assert(tp->tp_flags & Py_TPFLAGS_MANAGED_DICT);
```
Comment (markshannon) [inline Python/bytecodes.c]:
```suggestion
            assert(Py_TYPE(owner)->tp_flags & Py_TPFLAGS_MANAGED_DICT);
```

**Review by markshannon — CHANGES_REQUESTED** [review]
I'm hoping to remove `STORE_ATTR_WITH_HINT` once almost all instance have inline values.
Until then, this seems like a worthwhile improvement.

One question and a couple of suggestions.

**Review by Fidget-Spinner — APPROVED** [review]

**Pull request merged**